Index: control
===================================================================
--- control	(revision 2213)
+++ control	(working copy)
@@ -4,21 +4,21 @@
 Maintainer: Werner Mayer <wmayer@users.sourceforge.net>
 Homepage: http://sourceforge.net/projects/free-cad
 Build-Depends: debhelper (>= 5), autotools-dev, libc6-dev (>= 2.1.3),
- libstdc++6, libboost-dev, libboost-date-time-dev, libboost-filesystem-dev,
- libboost-graph-dev, libboost-iostreams-dev, libboost-program-options-dev,
- libboost-regex-dev, libboost-serialization-dev, libboost-signals-dev,
- zlib1g-dev, libxerces27-dev | libxerces-c2-dev,
+ libstdc++6, libboost1.37-dev, libboost-date-time1.37-dev, libboost-filesystem1.37-dev,
+ libboost-graph1.37-dev, libboost-iostreams1.37-dev, libboost-program-options1.37-dev,
+ libboost-regex1.37-dev, libboost-serialization1.37-dev, libboost-signals1.37-dev,
+ zlib1g-dev, libxerces-c28 | libxerces-c2-dev,
  libxt-dev, libxmu-dev, libxi-dev, libx11-dev, libxext-dev,
- libqt4-dev, libsoqt4-dev, libcoin40-dev, libgl1-mesa-dev,
+ libqt4-dev, libsoqt4-dev, libcoin60-dev, libgl1-mesa-dev,
  python2.5-dev, python, python-central (>= 0.5.6),
- libgts-dev, libcv-dev, libopencascade-dev
+ libgts-dev, libcv-dev, libopencascade-foundation-dev, libopencascade-modeling-dev
 Standards-Version: 3.7.3
 XS-Python-Version: current
 
 Package: freecad
 Architecture: any
 Section: science
-Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}, python
+Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}, python, python-pivy
 XB-Python-Version: ${python:Versions}
 Conflicts: freecad (<= 0.6.472-1)
 Suggests: gnochm | kchmviewer | kchmviewer-nokde | xchm, python-opencv
