Index: debian/control
===================================================================
--- debian/control	(Revision 1879)
+++ debian/control	(Arbeitskopie)
@@ -4,15 +4,15 @@
 Maintainer: Werner Mayer <wmayer@users.sourceforge.net>
 Homepage: http://sourceforge.net/projects/free-cad
 Build-Depends: debhelper (>= 5), autotools-dev, libc6-dev (>= 2.1.3),
- libstdc++6, libboost-dev, libboost-date-time-dev, libboost-filesystem-dev,
- libboost-graph-dev, libboost-iostreams-dev, libboost-program-options-dev,
- libboost-regex-dev, libboost-serialization-dev, libboost-signals-dev,
- zlib1g-dev, libxerces27-dev | libxerces-c2-dev,
+ libstdc++6, libboost1.35-dev, libboost-date-time1.35-dev, libboost-filesystem1.35-dev,
+ libboost-graph1.35-dev, libboost-iostreams1.35-dev, libboost-program-options1.35-dev,
+ libboost-regex1.35-dev, libboost-serialization1.35-dev, libboost-signals1.35-dev,
+ libboost-system1.35-dev, zlib1g-dev, libxerces27-dev | libxerces-c2-dev,
  libxt-dev, libxmu-dev, libxi-dev, libx11-dev, libxext-dev,
  libqt4-dev, libsoqt4-dev, libcoin40-dev, libgl1-mesa-dev,
  python2.5-dev, python, python-central (>= 0.5.6),
  libgts-dev, libcv-dev, libopencascade-dev
-Standards-Version: 3.7.3
+Standards-Version: 3.8.0
 XS-Python-Version: current
 
 Package: freecad
