Index: debian/rules
===================================================================
--- debian/rules	(revision 1523)
+++ debian/rules	(working copy)
@@ -11,6 +11,10 @@
 
 MODULES = Part Mesh Points Raytracing Image Drawing Test
 
+# Preliminary only as long as no official debian package for
+# OpenCascade is available
+DESTOCCDIR=$(shell pwd)/debian/freecad/usr/lib/freecad/lib
+
 # These are used for cross-compiling and for saving the configure script
 # from having to guess our platform (since we know it already)
 DEB_HOST_GNU_TYPE   ?= $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
@@ -85,6 +89,33 @@
 	dh_install -pfreecad debian/tmp/freecad/usr/lib/freecad/lib/FreeCAD.so usr/lib/freecad/lib
 	dh_install -pfreecad debian/tmp/freecad/usr/lib/freecad/lib/lib*.so.* usr/lib/freecad/lib
 
+# Preliminary only as long as no official debian package for
+# OpenCascade is available
+	install -m777 /usr/lib/libTKernel-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKMath-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKG2d-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKG3d-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKGeomBase-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKBRep-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKGeomAlgo-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKTopAlgo-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKPrim-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKBO-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKBool-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKShHealing-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKXSBase-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKSTEPBase-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKSTEPAttr-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKSTEP209-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKSTEP-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKFillet-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKOffset-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKIGES-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKMesh-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKHLR-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libTKSTL-6.2.so $(DESTOCCDIR)
+	install -m777 /usr/lib/libSoQt4.so.20 $(DESTOCCDIR)
+
 # install the modules
 	$(foreach MODULE,$(MODULES), \
 	dh_install -pfreecad debian/tmp/freecad/usr/lib/freecad/Mod/$(MODULE)/$(MODULE)*.so usr/lib/freecad/Mod/$(MODULE); \
Index: debian/shlibs.local
===================================================================
--- debian/shlibs.local	(revision 0)
+++ debian/shlibs.local	(revision 0)
@@ -0,0 +1,24 @@
+libTKernel 6.2
+libTKMath 6.2
+libTKG2d 6.2
+libTKG3d 6.2
+libTKGeomBase 6.2
+libTKBRep 6.2
+libTKGeomAlgo 6.2
+libTKTopAlgo 6.2
+libTKPrim 6.2
+libTKBO 6.2
+libTKBool 6.2
+libTKShHealing 6.2
+libTKXSBase 6.2
+libTKSTEPAttr 6.2
+libTKSTEPBase 6.2
+libTKSTEP209 6.2
+libTKSTEP 6.2
+libTKFillet 6.2
+libTKOffset 6.2
+libTKIGES 6.2
+libTKMesh 6.2
+libTKHLR 6.2
+libTKSTL 6.2
+libSoQt4 20
