--- pivy-0.3.0.orig/interfaces/soqt.i
+++ pivy-0.3.0/interfaces/soqt.i
@@ -78,7 +78,7 @@
 %include pivy_common_typemaps.i
 
 /* import the pivy main interface file */
-%import coin.i
+%import(module="pivy") coin.i
 
 /* typemaps to bridge against PyQt */
 %typemap(out) QEvent * {
--- pivy-0.3.0.orig/debian/changelog
+++ pivy-0.3.0/debian/changelog
@@ -0,0 +1,5 @@
+pivy (0.3.0-1) unstable; urgency=low
+
+  * Initial release (Closes: #504617)
+
+ -- Teemu Ikonen <tpikonen@gmail.com>  Thu, 13 Nov 2008 15:31:51 +0100
--- pivy-0.3.0.orig/debian/rules
+++ pivy-0.3.0/debian/rules
@@ -0,0 +1,62 @@
+#!/usr/bin/make -f
+
+CFLAGS = -Wall -g
+
+build: build-stamp
+build-stamp:
+	dh_testdir
+	python setup.py build
+	touch $@
+
+clean:
+	dh_testdir
+	dh_testroot
+	rm -f build-stamp install-stamp
+	python setup.py clean
+	dh_clean
+
+install: build install-stamp
+install-stamp:
+	dh_testdir
+	dh_testroot
+	dh_clean -k
+	dh_installdirs
+	python setup.py install --no-compile --root debian/python-pivy
+	#dh install --after dh_installdirs
+	touch install-stamp
+
+binary-indep: build install
+	#dh binary-indep
+
+binary-arch: build install
+	#dh binary-arch
+	dh_testdir
+	dh_testroot
+	dh_installchangelogs 
+	dh_installdocs
+	dh_installexamples
+#	dh_install
+#	dh_installmenu
+#	dh_installdebconf	
+#	dh_installlogrotate
+#	dh_installemacsen
+#	dh_installpam
+#	dh_installmime
+#	dh_desktop
+#	dh_installinit
+#	dh_installcron
+#	dh_installinfo
+	dh_installman
+	dh_link
+	dh_strip
+	dh_compress
+	dh_fixperms
+	dh_pycentral
+	dh_installdeb
+	dh_shlibdeps
+	dh_gencontrol
+	dh_md5sums
+	dh_builddeb
+	
+binary: binary-indep binary-arch
+.PHONY: build clean binary-indep binary-arch binary install
--- pivy-0.3.0.orig/debian/control
+++ pivy-0.3.0/debian/control
@@ -0,0 +1,25 @@
+Source: pivy
+Section: python
+Priority: extra
+Maintainer: Teemu Ikonen <tpikonen@gmail.com>
+Build-Depends: debhelper (>= 6), python-support, python-all-dev, 
+ libsoqt4-dev, libcoin40-dev, libsimage-dev, swig
+Standards-Version: 3.8.0
+Homepage: http://pivy.coin3d.org/
+
+Package: python-pivy
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}
+Provides: ${python:Provides}
+Description: Coin binding for Python
+ Pivy is a Coin binding for Python. Coin is a high-level 3D graphics library
+ with a C++ Application Programming Interface. Coin uses scene-graph data
+ structures to render real-time graphics suitable for mostly all kinds of
+ scientific and engineering visualization applications.
+ Pivy allows:
+ .
+    * Development of Coin applications and extensions in Python
+    * Interactive modification of Coin programs from within the
+      Python interpreter at runtime
+    * Incorporation of Scripting Nodes into the scene graph which
+      are capable of executing Python code and callbacks
--- pivy-0.3.0.orig/debian/compat
+++ pivy-0.3.0/debian/compat
@@ -0,0 +1 @@
+6
--- pivy-0.3.0.orig/debian/pyversions
+++ pivy-0.3.0/debian/pyversions
@@ -0,0 +1 @@
+2.4-
--- pivy-0.3.0.orig/debian/pycompat
+++ pivy-0.3.0/debian/pycompat
@@ -0,0 +1 @@
+2
--- pivy-0.3.0.orig/debian/docs
+++ pivy-0.3.0/debian/docs
@@ -0,0 +1,6 @@
+NEWS
+README
+HACKING
+THANKS
+docs/ruby-inventor.txt
+
--- pivy-0.3.0.orig/debian/watch
+++ pivy-0.3.0/debian/watch
@@ -0,0 +1,10 @@
+# See uscan(1) for format
+
+# Compulsory line, this is a version 3 file
+version=3
+
+# <Webpage URL> <string match>
+http://pivy.coin3d.org/download/pivy Pivy-(.*)\.tar\.gz
+
+# Uncomment to examine a Webserver directory
+#http://www.example.com/pub/pivy-(.*)\.tar\.gz
--- pivy-0.3.0.orig/debian/copyright
+++ pivy-0.3.0/debian/copyright
@@ -0,0 +1,58 @@
+Format-Specification: http://wiki.debian.org/Proposals/CopyrightFormat
+Upstream-Author: Tamer Fahmy <tamer@tammura.at>
+Packaged-By: Teemu Ikonen <tpikonen@gmail.com>
+Original-Source-Location: http://pivy.coin3d.org/download/pivy/releases/0.3.0/Pivy-0.3.0.tar.gz
+
+Files: *
+Copyright: 2002-2005 Tamer Fahmy <tamer@tammura.at>
+License: BSD-3
+ Redistribution and use in source and binary forms, with or without
+ modification, are permitted provided that the following conditions are
+ met:
+  * Redistributions of source code must retain the above copyright
+    notice, this list of conditions and the following disclaimer.
+  * Redistributions in binary form must reproduce the above copyright
+    notice, this list of conditions and the following disclaimer in
+    the documentation and/or other materials provided with the
+    distribution.
+  * Neither the name of the copyright holder nor the names of its
+    contributors may be used to endorse or promote products derived
+    from this software without specific prior written permission.
+ .
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+Files: scons/*
+Copyright: 2001, 2002, 2003, 2004 The SCons Foundation
+License: other
+ Permission is hereby granted, free of charge, to any person obtaining
+ a copy of this software and associated documentation files (the
+ "Software"), to deal in the Software without restriction, including
+ without limitation the rights to use, copy, modify, merge, publish,
+ distribute, sublicense, and/or sell copies of the Software, and to
+ permit persons to whom the Software is furnished to do so, subject to
+ the following conditions:
+ .
+ The above copyright notice and this permission notice shall be included
+ in all copies or substantial portions of the Software.
+ .
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
+ KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+Files: debian/*
+Copyright: 2008 Teemu Ikonen <tpikonen@gmail.com>
+License: BSD-3
