--- pythonocc-0.3.0.orig/src/wrapper/environment.py
+++ pythonocc-0.3.0/src/wrapper/environment.py
@@ -126,8 +126,8 @@
         OCC_INC = os.path.join(OCC_ROOT,'inc')
         OCC_LIB = os.path.join(OCC_ROOT,'lib')
     else:
-        OCC_INC = '/usr/local/inc' # Ubuntu 8.04 Hardy default installation path for headers
-        OCC_LIB = '/usr/local/lib' # Ubuntu 8.04 Hardy default installation path for libraries
+        OCC_INC = '/usr/include/opencascade' # Ubuntu 8.04 Hardy default installation path for headers
+        OCC_LIB = '/usr/lib' # Ubuntu 8.04 Hardy default installation path for libraries
     SALOME_GEOM_LIB = '/opt/SalomeGeometry/lib'
     GCC_XML_PATH = '/usr/bin' # Ubuntu 8.04 Hardy installation path for gccxml
     PYGCCXML_DEFINES = ['HAVE_CONFIG_H','HAVE_LIMITS_H','CSFDB','OCC_CONVERT_SIGNALS','LIN','LININTEL','_GNU_SOURCE=1']
--- pythonocc-0.3.0.orig/src/setup.py
+++ pythonocc-0.3.0/src/setup.py
@@ -172,18 +172,18 @@
 #
 # OpenCascade libs
 #
-libraries = ['BinLPlugin', 'BinPlugin', 'BinXCAFPlugin', 'FWOSPlugin', 'mscmd', 'PTKernel',\
+libraries = ['BinLPlugin', 'BinPlugin', 'BinXCAFPlugin', 'FWOSPlugin', 'PTKernel',\
              'StdLPlugin', 'StdPlugin', 'TKAdvTools', 'TKBin', 'TKBinL', 'TKBinTObj', 'TKBinXCAF',\
-             'TKBO', 'TKBool', 'TKBRep', 'TKCAF', 'TKCDF', 'TKCDLFront', 'TKCPPClient', 'TKCPPExt',\
-             'TKCPPIntExt', 'TKCPPJini', 'TKCSFDBSchema', 'TKDCAF', 'TKDraw', 'TKernel',\
-             'TKFeat', 'TKFillet', 'TKG2d', 'TKG3d', 'TKGeomAlgo', 'TKGeomBase', 'TKHLR', 'TKIDLFront',\
+             'TKBO', 'TKBool', 'TKBRep', 'TKCAF', 'TKCDF',\
+             'TKernel',\
+             'TKFeat', 'TKFillet', 'TKG2d', 'TKG3d', 'TKGeomAlgo', 'TKGeomBase', 'TKHLR',\
              'TKIGES', 'TKjcas','TKLCAF', 'TKMath', 'TKMesh', 'TKMeshVS', 'TKNIS', 'TKOffset',\
              'TKOpenGl', 'TKPCAF', 'TKPLCAF', 'TKPrim', 'TKPShape', 'TKService', 'TKShapeSchema',\
              'TKShHealing', 'TKStdLSchema', 'TKStdSchema', 'TKSTEP', 'TKSTEP209', 'TKSTEPAttr',\
-             'TKSTEPBase', 'TKSTL', 'TKTCPPExt', 'TKTObj', 'TKTObjDRAW', 'TKTopAlgo', 'TKTopTest',\
-             'TKV2d', 'TKV3d', 'TKViewerTest', 'TKVRML', 'TKWOK', 'TKWOKTcl', 'TKXCAF', 'TKXCAFSchema',\
-             'TKXDEDRAW', 'TKXDEIGES', 'TKXDESTEP', 'TKXMesh', 'TKXml', 'TKXmlL', 'TKXmlTObj',\
-             'TKXmlXCAF', 'TKXSBase', 'TKXSDRAW', 'XCAFPlugin',\
+             'TKSTEPBase', 'TKSTL', 'TKTObj', 'TKTObjDRAW', 'TKTopAlgo',\
+             'TKV2d', 'TKV3d', 'TKVRML', 'TKXCAF', 'TKXCAFSchema',\
+             'TKXDEIGES', 'TKXDESTEP', 'TKXMesh', 'TKXml', 'TKXmlL', 'TKXmlTObj',\
+             'TKXmlXCAF', 'TKXSBase',\
              'XmlLPlugin', 'XmlPlugin', 'XmlXCAFPlugin']
 # Find the lib in OCC_LIB path and add it to the LIBS list
 LIBS = []
--- pythonocc-0.3.0.orig/debian/changelog
+++ pythonocc-0.3.0/debian/changelog
@@ -0,0 +1,6 @@
+pythonocc (0.3.0-1) unstable; urgency=low
+
+  * Initial release
+
+ -- Werner Mayer <wmayer@users.sourceforge.net>  Mon, 24 Aug 2009 12:41:48 +0200
+
--- pythonocc-0.3.0.orig/debian/rules
+++ pythonocc-0.3.0/debian/rules
@@ -0,0 +1,80 @@
+#!/usr/bin/make -f
+
+CFLAGS = -Wall -g
+
+configure:
+	chmod 755 src/addons/Display/OCCViewer.py
+	chmod 755 src/addons/Display/wxDisplay.py
+	chmod 755 src/addons/Display/wxSamplesGui.py
+	chmod 755 src/addons/Display/wxSamplesGui2d.py
+	chmod 755 src/addons/Utils/DataExchange/STL.py
+	chmod 755 src/addons/Utils/DataExchange/STEP.py
+	chmod 755 src/addons/Utils/Common.py
+	chmod 755 src/addons/Utils/Image.py
+	#chmod 644 src/AUTHORS
+
+build: build-stamp
+build-stamp: configure
+	dh_testdir
+	cd src && python setup.py build
+	touch $@
+
+clean:
+	dh_testdir
+	dh_testroot
+	rm -f build-stamp install-stamp
+	cd src && python setup.py clean
+	dh_clean
+	rm -rf src/build
+	rm -rf src/OCC
+	find -name '*.pyc' | xargs rm -rf
+	cd src/wrapper && find -name '*.pyc' | xargs rm -rf
+	cd src/wrapper/Misc && find -name '*_wrap.cpp' | xargs rm -rf
+	cd src/wrapper/SWIG/linux_darwin && find -name '*_wrap.cpp' | xargs rm -rf
+	cd src/wrapper/Visualization && find -name '*_wrap.cpp' | xargs rm -rf
+
+install: build install-stamp
+install-stamp:
+	dh_testdir
+	dh_testroot
+	dh_clean -k
+	dh_installdirs
+	cd src && python setup.py install --no-compile --root ../debian/python-occ
+	#dh install --after dh_installdirs
+	touch install-stamp
+
+binary-indep: build install
+	#dh binary-indep
+
+binary-arch: build install
+	#dh binary-arch
+	dh_testdir
+	dh_testroot
+	dh_installchangelogs 
+	dh_installdocs
+	dh_installexamples
+#	dh_install
+#	dh_installmenu
+#	dh_installdebconf	
+#	dh_installlogrotate
+#	dh_installemacsen
+#	dh_installpam
+#	dh_installmime
+#	dh_desktop
+#	dh_installinit
+#	dh_installcron
+#	dh_installinfo
+	dh_installman
+	dh_link
+	dh_strip
+	dh_compress
+	dh_fixperms
+	dh_pycentral
+	dh_installdeb
+	dh_shlibdeps
+	dh_gencontrol
+	dh_md5sums
+	dh_builddeb
+
+binary: binary-indep binary-arch
+.PHONY: build clean binary-indep binary-arch binary install
--- pythonocc-0.3.0.orig/debian/control
+++ pythonocc-0.3.0/debian/control
@@ -0,0 +1,20 @@
+Source: pythonocc
+Section: python
+Priority: extra
+Maintainer: Werner Mayer <wmayer@users.sourceforge.net>
+Build-Depends: debhelper (>= 6), python, python-central, python-support, swig, libopencascade-dev, salomegeom-dev
+Standards-Version: 3.7.3
+XS-Python-Version: current
+Homepage: http://pythonocc.org/
+
+Package: python-occ
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends}, ${python:Depends}
+XB-Python-Version: ${python:Versions}
+Provides: ${python:Provides}
+Description: A framework for agile CAD development
+ PythonOCC is a Python wrapper module for the
+ OpenCascade library. It contains python functions and classes
+ that will allow you to fully utilitize the OpenCascade library.
+ This version is built against OpenCascade 6.3.0.
+
--- pythonocc-0.3.0.orig/debian/compat
+++ pythonocc-0.3.0/debian/compat
@@ -0,0 +1 @@
+6
--- pythonocc-0.3.0.orig/debian/pyversions
+++ pythonocc-0.3.0/debian/pyversions
@@ -0,0 +1 @@
+2.4-
--- pythonocc-0.3.0.orig/debian/pycompat
+++ pythonocc-0.3.0/debian/pycompat
@@ -0,0 +1 @@
+2
--- pythonocc-0.3.0.orig/debian/copyright
+++ pythonocc-0.3.0/debian/copyright
@@ -0,0 +1,7 @@
+Files: *
+Copyright: 2008 Thomas Paviot <thomas.paviot@free.fr> 
+License: GPL-3
+ .
+ On Debian systems, the complete text of the GNU General Public 
+ License version 3 can be found in `/usr/share/common-licenses/GPL-3'.
+
