--- salomegeom-5.1.1.orig/adm/lin/Makefile.am
+++ salomegeom-5.1.1/adm/lin/Makefile.am
@@ -2,10 +2,10 @@
 # geomdir = ./
 
 # copy some files
-targetdir = $(prefix)
-dist_target_DATA = LICENCE.lgpl
+#targetdir = $(datadir)
+#dist_target_DATA = LICENCE.lgpl
 
-resourcedir = $(prefix)/resources
+resourcedir = $(datadir)/resources
 dist_resource_DATA = $(srcdir)/resources/GUID.txt $(srcdir)/resources/ImportExport $(srcdir)/resources/Resources $(srcdir)/resources/ShHealing
 
 
@@ -16,7 +16,7 @@
 AM_CPPFLAGS = -I$(CASROOT)/inc -I$(srcdir)/$(geomdir)inc/ -I$(srcdir)/$(geomdir)src/GEOMAlgo -I$(srcdir)/$(geomdir)src/BlockFix -I$(srcdir)/$(geomdir)src/GEOM -I$(srcdir)/$(geomdir)src/NMTTools -I$(srcdir)/$(geomdir)src/NMTDS
 
 # set library names
-lib_LTLIBRARIES = libExchangeCSFDB.la libArchimede.la libNMTDS.la libNMTTools.la libGEOMAlgo.la libGEOMImpl.la libGEOM.la libExchangeBREP.la libExchangeIGES.la libExchangeSTEP.la libExchangeSTL.la libExchangeVRML.la libPartition.la libShHealOper.la libSketcher.la
+lib_LTLIBRARIES = libExchangeCSFDB.la libArchimede.la libNMTDS.la libNMTTools.la libGEOMAlgo.la libGEOM.la libShHealOper.la libExchangeBREP.la libExchangeIGES.la libExchangeSTEP.la libExchangeSTL.la libExchangeVRML.la libPartition.la libSketcher.la libGEOMImpl.la
 
 # set global linker flags
 AM_LDFLAGS = -version-info 1:0:0
@@ -24,24 +24,24 @@
 uninstall-local:
 	-rm -rf $(DESTDIR)$(libdir)
 	-rm -rf $(DESTDIR)$(includedir)
-	-rm -rf $(DESTDIR)$(prefix)/resources
-	-rm $(DESTDIR)$(prefix)/env_geom.sh
+	-rm -rf $(DESTDIR)$(datadir)/resources
+	-rm $(DESTDIR)$(datadir)/env_geom.sh
 
 install-exec-local:
-	test -z "$(DESTDIR)$(prefix)" || $(mkdir_p) "$(DESTDIR)$(prefix)"
-	@echo "#!/bin/bash" 											> $(DESTDIR)$(prefix)/env_geom.sh
-	@echo "export GEOMROOT="$(prefix) 								>> $(DESTDIR)$(prefix)/env_geom.sh
-	@echo "export CSF_ResourcesDefaults="$(prefix)/resources 		>> $(DESTDIR)$(prefix)/env_geom.sh
-	@echo "if [ -z "LD_LIBRARY_PATH" ];" 							>> $(DESTDIR)$(prefix)/env_geom.sh
-	@echo "then LD_LIBRARY_PATH="$(libdir)";" 						>> $(DESTDIR)$(prefix)/env_geom.sh
-	@echo "else LD_LIBRARY_PATH="$(libdir)":\$$LD_LIBRARY_PATH;" 	>> $(DESTDIR)$(prefix)/env_geom.sh 
-	@echo "fi" 														>> $(DESTDIR)$(prefix)/env_geom.sh 
-	chmod a+x $(DESTDIR)$(prefix)/env_geom.sh
-	test -z "$(DESTDIR)$(prefix)/resources" || $(mkdir_p) "$(DESTDIR)$(prefix)/resources"
-	cp $(srcdir)/resources/GUID.txt "$(DESTDIR)$(prefix)/resources"
-	cp $(srcdir)/resources/ImportExport "$(DESTDIR)$(prefix)/resources"
-	cp $(srcdir)/resources/Resources "$(DESTDIR)$(prefix)/resources"
-	cp $(srcdir)/resources/ShHealing "$(DESTDIR)$(prefix)/resources"
+	test -z "$(DESTDIR)$(datadir)" || $(mkdir_p) "$(DESTDIR)$(datadir)"
+	@echo "#!/bin/bash" 											> $(DESTDIR)$(datadir)/env_geom.sh
+	@echo "export GEOMROOT="$(prefix) 								>> $(DESTDIR)$(datadir)/env_geom.sh
+	@echo "export CSF_ResourcesDefaults="$(datadir)/resources 		>> $(DESTDIR)$(datadir)/env_geom.sh
+	@echo "if [ -z "LD_LIBRARY_PATH" ];" 							>> $(DESTDIR)$(datadir)/env_geom.sh
+	@echo "then LD_LIBRARY_PATH="$(libdir)";" 						>> $(DESTDIR)$(datadir)/env_geom.sh
+	@echo "else LD_LIBRARY_PATH="$(libdir)":\$$LD_LIBRARY_PATH;" 	>> $(DESTDIR)$(datadir)/env_geom.sh 
+	@echo "fi" 														>> $(DESTDIR)$(datadir)/env_geom.sh 
+	chmod a+x $(DESTDIR)$(datadir)/env_geom.sh
+	test -z "$(DESTDIR)$(datadir)/resources" || $(mkdir_p) "$(DESTDIR)$(datadir)/resources"
+	cp $(srcdir)/resources/GUID.txt "$(DESTDIR)$(datadir)/resources"
+	cp $(srcdir)/resources/ImportExport "$(DESTDIR)$(datadir)/resources"
+	cp $(srcdir)/resources/Resources "$(DESTDIR)$(datadir)/resources"
+	cp $(srcdir)/resources/ShHealing "$(DESTDIR)$(datadir)/resources"
 
 # set ARCHIMDE sources
 libArchimede_la_SOURCES =  $(srcdir)/$(geomdir)src/Archimede/Archimede_VolumeSection.cpp
@@ -582,6 +582,26 @@
 # libShHealOper_la_LDFLAGS        = -version-info 0:0:0
 # libSketcher_la_LDFLAGS          = -version-info 0:0:0
 
+libArchimede_la_LIBADD        = -lTKMath -lTKernel -lTKBRep -lTKGeomAlgo -lTKGeomBase -lTKG3d -lTKMesh
+libExchangeBREP_la_LIBADD     = -lTKMath -lTKernel -lTKBRep
+libExchangeCSFDB_la_LIBADD    = -lTKMath -lTKernel -lTKBRep -lTKPShape -lTKShapeSchema -lPTKernel
+libExchangeIGES_la_LIBADD     = -lTKMath -lTKernel -lTKXSBase -lTKBRep -lTKIGES
+libExchangeSTEP_la_LIBADD     = -lTKMath -lTKernel -lTKXSBase -lTKBRep -lTKSTEP
+libExchangeSTL_la_LIBADD      = -lTKernel -lTKSTL
+libExchangeVRML_la_LIBADD     = -lTKernel -lTKVRML
+libNMTDS_la_LIBADD            = -lTKMath -lTKernel -lTKBRep -lTKBO
+libNMTTools_la_LIBADD         = -lTKMath -lTKernel -lTKBRep -lTKBO -lTKG2d -lTKG3d -lTKGeomAlgo -lTKGeomBase -lTKTopAlgo -lNMTDS
+libGEOMAlgo_la_LIBADD         = -lTKMath -lTKernel -lTKBRep -lTKBool -lTKBO -lTKG2d -lTKG3d -lTKGeomAlgo \
+								-lTKGeomBase -lTKTopAlgo -lTKMesh -lTKShHealing -lNMTDS -lNMTTools
+libGEOM_la_LIBADD             = -lTKMath -lTKernel -lTKXSBase -lTKBRep -lTKCAF -lTKCDF -lTKLCAF -lTKXSBase \
+								-lTKAdvTools -lTKG2d
+libGEOMImpl_la_LIBADD         = -lTKMath -lTKernel -lTKBRep -lTKTopAlgo -lTKCAF -lTKOffset -lTKPrim -lTKShHealing \
+								-lTKBool -lTKBO -lTKMesh -lTKG2d -lTKG3d -lTKFillet -lTKGeomAlgo -lTKGeomBase -lTKLCAF \
+								-lGEOMAlgo -lGEOM -lArchimede -lShHealOper -lSketcher -ldl
+libPartition_la_LIBADD        = -lTKMath -lTKernel -lTKBRep -lTKBool -lTKG2d -lTKG3d -lTKTopAlgo -lTKGeomAlgo -lTKGeomBase -lTKOffset
+libShHealOper_la_LIBADD       = -lTKMath -lTKernel -lTKBRep -lTKBool -lTKShHealing -lTKTopAlgo -lTKG2d -lTKG3d -lTKGeomAlgo -lTKGeomBase -lTKLCAF
+libSketcher_la_LIBADD         = -lTKMath -lTKernel -lTKBRep -lTKTopAlgo -lTKG2d -lTKGeomAlgo
+
 # include files to be copied to installdir/inc
 include_HEADERS =  $(srcdir)/$(geomdir)inc/Archimede_VolumeSection.hxx \
 $(srcdir)/$(geomdir)inc/BlockFix_BlockFixAPI.hxx \
@@ -921,7 +941,6 @@
 $(srcdir)/$(geomdir)inc/ShHealOper_RemoveInternalWires.hxx \
 $(srcdir)/$(geomdir)inc/ShHealOper_Sewing.hxx \
 $(srcdir)/$(geomdir)inc/ShHealOper_ShapeProcess.hxx \
-$(srcdir)/$(geomdir)inc/ShHealOper_SpiltCurve2d.hxx \
 $(srcdir)/$(geomdir)inc/ShHealOper_SplitCurve2d.hxx \
 $(srcdir)/$(geomdir)inc/ShHealOper_SplitCurve3d.hxx \
 $(srcdir)/$(geomdir)inc/ShHealOper_Tool.hxx \
--- salomegeom-5.1.1.orig/debian/changelog
+++ salomegeom-5.1.1/debian/changelog
@@ -0,0 +1,6 @@
+salomegeom (5.1.1-6) unstable; urgency=low
+
+  * Initial release
+
+ -- Werner Mayer <wmayer@users.sourceforge.net>  Sun, 23 Aug 2009 17:12:19 +0200
+
--- salomegeom-5.1.1.orig/debian/rules
+++ salomegeom-5.1.1/debian/rules
@@ -0,0 +1,129 @@
+#!/usr/bin/make -f
+# -*- makefile -*-
+# Sample debian/rules that uses debhelper.
+# This file was originally written by Joey Hess and Craig Small.
+# As a special exception, when this file is copied by dh-make into a
+# dh-make output file, you may use that output file without restriction.
+# This special exception was added by Craig Small in version 0.37 of dh-make.
+
+# Uncomment this to turn on verbose mode.
+#export DH_VERBOSE=1
+
+# These are used for cross-compiling and for saving the configure script
+# from having to guess our platform (since we know it already)
+DEB_HOST_GNU_TYPE   ?= $(shell dpkg-architecture -qDEB_HOST_GNU_TYPE)
+DEB_BUILD_GNU_TYPE  ?= $(shell dpkg-architecture -qDEB_BUILD_GNU_TYPE)
+BUILD_DIR = adm/lin
+
+CFLAGS = -Wall -g
+
+ifneq (,$(findstring noopt,$(DEB_BUILD_OPTIONS)))
+	CFLAGS += -O0
+else
+	CFLAGS += -O2
+endif
+
+configure:
+	chmod u-x resources/GUID.txt
+	chmod u-x resources/ImportExport
+	chmod u-x resources/Resources
+	chmod u-x resources/ShHealing
+	chmod u-x resources/ShHealingFullSet
+	cd $(BUILD_DIR) && \
+	autoreconf --install
+
+config.status: configure
+	dh_testdir
+# Add here commands to configure the package.
+	cd $(BUILD_DIR) && \
+	./configure --host=$(DEB_HOST_GNU_TYPE) --build=$(DEB_BUILD_GNU_TYPE) \
+	--prefix=/usr --includedir=/usr/include/salomegeom \
+	--libdir=/usr/lib --datadir=/usr/share/salomegeom \
+	LDFLAGS="-Wl,-z,defs -L/usr/lib" \
+	CXXFLAGS="$(CXXFLAGS) -I/usr/include/opencascade" 
+
+build: build-links build-stamp
+
+build-links:
+	dh_testdir
+	cd $(BUILD_DIR) && \
+	ln -s ../../inc inc && \
+	ln -s ../../resources resources && \
+	ln -s ../../src src
+
+build-stamp:  config.status
+	dh_testdir
+
+# Add here commands to compile the package.
+	cd $(BUILD_DIR) && $(MAKE)
+	touch $@
+
+# Add here commands to clean up after the build process.
+clean:
+	dh_testdir
+	dh_testroot
+	dh_clean
+# We always get some strange symlinks in these directories which we must remove first
+	rm -f inc/inc
+	rm -f src/src
+	rm -f resources/resources
+	rm -f build-stamp
+	cd $(BUILD_DIR) && \
+	rm -f configure && \
+	rm -f inc resources src && \
+	find -name 'Makefile.in' | xargs rm -f && \
+	find -name 'Makefile' | xargs rm -f && \
+	find -name '*.o' | xargs rm -f && \
+	find -name '*.la' | xargs rm -f && \
+	find -name '*.lo' | xargs rm -f && \
+	find -name '*.deps' | xargs rm -rf && \
+	find -name '*.libs' | xargs rm -rf	 && \
+	rm -f stamp-h1 aclocal.m4 config.guess config.sub depcomp install-sh && \
+	rm -f ltmain.sh missing config.log libtool && \
+	rm -f config.h config.h.in config.status && \
+	rm -rf build-aux autom4te.cache
+	
+install: build-stamp
+	dh_testdir
+	dh_testroot
+	dh_clean -k 
+	dh_installdirs
+
+# Add here commands to install the package into debian/salomegeom.
+	cd $(BUILD_DIR) && $(MAKE) install DESTDIR=$(CURDIR)/debian/tmp/salomegeom
+
+# install the core system
+	dh_install -psalomegeom debian/tmp/salomegeom/usr/share/* usr/share/
+	dh_install -psalomegeom debian/tmp/salomegeom/usr/lib/lib*.so.* usr/lib
+	dh_install -psalomegeom-dev debian/tmp/salomegeom/usr/lib/lib*.so usr/lib
+	dh_install -psalomegeom-dev debian/tmp/salomegeom/usr/lib/lib*.la usr/lib
+	dh_install -psalomegeom-dev debian/tmp/salomegeom/usr/lib/lib*.a  usr/lib
+	dh_install -psalomegeom-dev debian/tmp/salomegeom/usr/include/* usr/include
+
+# Build architecture-independent files here.
+binary-indep: build install
+# We have nothing to do by default.
+
+# Build architecture-dependent files here.
+binary-arch: build install
+	dh_testdir
+	dh_testroot
+	dh_installchangelogs
+	dh_installdocs
+	dh_installexamples
+	dh_desktop
+#	dh_installinfo
+	dh_link
+	dh_strip
+	dh_compress
+	dh_fixperms
+	dh_makeshlibs
+	dh_installdeb
+	dh_shlibdeps
+	dh_gencontrol
+	dh_md5sums
+	dh_builddeb
+
+binary: binary-indep binary-arch
+.PHONY: build clean binary-indep binary-arch binary install 
+
--- salomegeom-5.1.1.orig/debian/control
+++ salomegeom-5.1.1/debian/control
@@ -0,0 +1,26 @@
+Source: salomegeom
+Section: libs
+Priority: extra
+Maintainer: Werner Mayer <wmayer@users.sourceforge.net>
+Build-Depends: cdbs, debhelper (>= 5), libopencascade-dev
+Standards-Version: 3.7.3
+
+Package: salomegeom
+Architecture: any
+Depends: ${shlibs:Depends}, ${misc:Depends}
+Description: SALOME Geometry Module
+ A complete OpenCascade - OCAF based CAD framework.Note that this is not
+ the original SALOME GEOM project but an effort to create a standalone
+ framework based on the existing one from SALOME project, with the
+ addition of extra functionality.
+
+Package: salomegeom-dev
+Architecture: any
+Section: libdevel
+Depends: salomegeom (>= ${source:Version})
+Description: SALOME Geometry Module development files
+ A complete OpenCascade - OCAF based CAD framework.Note that this is not
+ the original SALOME GEOM project but an effort to create a standalone
+ framework based on the existing one from SALOME project, with the
+ addition of extra functionality.
+
--- salomegeom-5.1.1.orig/debian/compat
+++ salomegeom-5.1.1/debian/compat
@@ -0,0 +1 @@
+5
--- salomegeom-5.1.1.orig/debian/README.Debian
+++ salomegeom-5.1.1/debian/README.Debian
@@ -0,0 +1,6 @@
+salomegeom for Debian
+---------------------
+
+- First Debian package for SalomeGeom.
+
+ -- Werner Mayer <wmayer@users.sourceforge.net>  Sun, 23 Aug 2009 17:12:19 +0200
--- salomegeom-5.1.1.orig/debian/docs
+++ salomegeom-5.1.1/debian/docs
@@ -0,0 +1 @@
+README.LINUX
--- salomegeom-5.1.1.orig/debian/copyright
+++ salomegeom-5.1.1/debian/copyright
@@ -0,0 +1,16 @@
+This package was debianized by Werner Mayer <wmayer@users.sourceforge.net> on
+Sun, 23 Aug 2009 17:12:19 +0200.
+
+It was downloaded from https://salomegeometry.svn.sourceforge.net/svnroot/salomegeometry/tags/5.1.1_v6
+
+Upstream Author: 
+    Fotis Sioutis <sfotis@users.sourceforge.net>
+
+Copyright: 
+    Copyright (C) 2005  OPEN CASCADE, EADS/CCR, LIP6, CEA/DEN,
+    CEDRAT, EDF R&D, LEG, PRINCIPIA R&D, BUREAU VERITAS
+
+License: LGPL-2.1
+ On Debian systems, the complete text of the GNU Lesser General Public 
+ License version 2.1 can be found in `/usr/share/common-licenses/LGPL-2.1'.
+
